File: src/core/gui/DropDown.ts
import { disposable_listener, div, Icon } from "./dom";
import "./DropDown.css";
import { Property } from "../observable/property/Property";
import { Button, ButtonOptions } from "./Button";
import { Menu } from "./Menu";
import { Control } from "./Control";
import { Observable } from "../observable/Observable";
import { Emitter } from "../observable/Emitter";
import { emitter } from "../observable";

export type DropDownOptions<T> = ButtonOptions & {
    readonly text: string;
    readonly items: readonly T[] | Property<readonly T[]>;
    readonly to_label?: (element: T) => string;
};

export class DropDown<T> extends Control {
    readonly element = div({ className: "core_DropDown" });

    readonly chosen: Observable<T>;

    private readonly button: Button;
    private readonly menu: Menu<T>;
    private readonly _chosen: Emitter<T>;
    private just_opened: boolean;

    constructor(options: DropDownOptions<T>) {
        super(options);

        this.button = this.disposable(
            new Button({
                text: options.text,
                icon_left: options && options.icon_left,
                icon_right: Icon.TriangleDown,
            }),
        );
        this.menu = this.disposable(
            new Menu({
                items: options.items,
                to_label: options.to_label,
                related_element: this.element,
            }),
        );
        this.element.append(this.button.element, this.menu.element);

        this._chosen = emitter();
        this.chosen = this._chosen;

        this.just_opened = false;

        this.disposables(
            disposable_listener(this.button.element, "mousedown", this.button_mousedown, {
                capture: true,
            }),

            this.button.onmouseup.observe(this.button_mouseup),

            this.button.onkeydown.observe(this.button_keydown),

            this.menu.selected.observe(({ value }) => {
                if (value !== undefined) {
                    this._chosen.emit({ value });
                    this.menu.selected.val = undefined;
                }
            }),
        );

        this.finalize_construction();
    }

    protected set_enabled(enabled: boolean): void {
        super.set_enabled(enabled);
        this.button.enabled.val = enabled;
    }

    private button_mousedown = (): void => {
        this.just_opened = !this.menu.visible.val;
        this.menu.visible.val = true;
    };

    private button_mouseup = (): void => {
        if (this.just_opened) {
            this.menu.focus();
        } else {
            this.menu.visible.val = false;
        }

        this.just_opened = false;
    };

    private button_keydown = ({ value: evt }: { value: KeyboardEvent }): void => {
        if (evt.key === "Enter" || evt.key === " ") {
            evt.preventDefault();
            this.just_opened = !this.menu.visible.val;
            this.menu.visible.val = true;
            this.menu.focus();
            this.menu.hover_next();
        }
    };
}

File: webui/src/main/kotlin/world/phantasmal/webui/widgets/DropDown.kt
package world.phantasmal.webui.widgets

import org.w3c.dom.Node
import org.w3c.dom.events.KeyboardEvent
import org.w3c.dom.events.MouseEvent
import world.phantasmal.observable.cell.Cell
import world.phantasmal.observable.cell.list.emptyListCell
import world.phantasmal.observable.cell.mutableCell
import world.phantasmal.observable.cell.nullCell
import world.phantasmal.observable.cell.trueCell
import world.phantasmal.webui.dom.Icon
import world.phantasmal.webui.dom.div

class Dropdown<T : Any>(
    visible: Cell<Boolean> = trueCell(),
    enabled: Cell<Boolean> = trueCell(),
    tooltip: Cell<String?> = nullCell(),
    private val text: String? = null,
    private val iconLeft: Icon? = null,
    items: Cell<List<T>>? = null,
    private val itemToString: (T) -> String = Any::toString,
    private val onSelect: (T) -> Unit = {},
) : Control(visible, enabled, tooltip) {
    private val items: Cell<List<T>> = items ?: emptyListCell()

    private val menuVisible = mutableCell(false)

    private lateinit var menu: Menu<T>
    private var justOpened = false

    override fun Node.createElement() =
        div {
            className = "pw-dropdown"

            addWidget(Button(
                enabled = enabled,
                text = text,
                iconLeft = iconLeft,
                iconRight = Icon.TriangleDown,
                onMouseDown = ::onButtonMouseDown,
                onMouseUp = { onButtonMouseUp() },
                onKeyDown = ::onButtonKeyDown,
            ))
            menu = addWidget(Menu(
                visible = menuVisible,
                enabled = enabled,
                items = items,
                itemToString = itemToString,
                onSelect = ::select,
                onCancel = { menuVisible.value = false },
            ))
        }


    private fun onButtonMouseDown(e: MouseEvent) {
        e.stopPropagation()
        justOpened = !menuVisible.value
        menuVisible.value = true
    }

    private fun onButtonMouseUp() {
        if (justOpened) {
            menu.focus()
        } else {
            menuVisible.value = false
        }

        justOpened = false
    }

    private fun onButtonKeyDown(e: KeyboardEvent) {
        when (e.key) {
            "Enter", " " -> {
                e.preventDefault()
                e.stopPropagation()

                justOpened = !menuVisible.value
                menuVisible.value = true
                items.value.firstOrNull()?.let(menu::highlightItem)
                menu.focus()
            }
        }
    }

    private fun select(item: T) {
        menuVisible.value = false
        onSelect(item)
    }

    companion object {
        init {
            @Suppress("CssUnusedSymbol")
            // language=css
            style("""
                .pw-dropdown {
                    position: relative;
                }

                .pw-dropdown > .pw-menu {
                    top: 23px;
                    left: 0;
                    min-width: 100%;
                }
            """.trimIndent())
        }
    }
}
